infra/ops#1014 Move localized default settings into civicrm-core

mlutfy opened the pull request:
Overview
----------------------------------------

When someone installs CiviCRM in another language/locale than `en_US`, whether it's `en_GB` or `fr_CA`, CiviCRM will load default settings specific to that locale (such as currency, address format, date/time formats, default country, etc).

This was previously in civicrm-l10n, however, the l10n tar.gz is being deprecated.

https://lab.civicrm.org/infra/ops/-/issues/1014

Before
----------------------------------------

json files had to be installed manually.

After
----------------------------------------

No extra steps necessary.

Technical Details
--------------------------

I wasn't sure about using `$civicrm_root`, since this is called during installation, and I'm not sure what is available at that point. I copied an example I found while grepping around.

Comments
----------------------------------------

Needs a bit more testing.

Edit: available in CiviCRM 6.3.0.

h/t @cividesk

mlutfy committed:
infra/ops#1014 Move localized default settings into civicrm-core

This was previously in civicrm-l10n, however, the l10n tar.gz is being
deprecated.

Also includes:

* minor tweaks to the installer help texts and links
* allows installing sample data for non-US-English
* moves the init code to a setup plugin.
* download the l10n files before starting the installation (this makes it
  possible to localize the installer itself (which was possible back
  in the days of the l10n.tar.gz file).

totten commented on Civi/Core/LocalizationInitializer.php:
```diff
@@ -26,23 +26,19 @@ class LocalizationInitializer {
-    // initalization
+    // Get the corresponding settings file if any
+    global $civicrm_root;
+    $fileName = \CRM_Utils_File::addTrailingSlash($civicrm_root) . "settings/l10n/$seedLanguage/settings.default.json";
```
> I wasn't sure about using `$civicrm_root`, since this is called during installation, and I'm not sure what is available at that point. I copied an example I found while grepping around.

If it works here, then I wouldn't worry about it too much. ☮️ 

FWIW, this alternative starts working pretty early, and it feels a little prettier...

```php
$fileName = Civi::paths()->getPath("[civicrm.root]/settings/l10n/$seedLanguage/settings.default.json");
## N.B. `[civicrm.root]` has trailing-slash normalized already
```

But it's not worth an extra round of testing just for that. (It sounds like testing is the main thing.)

(We could try that another time -- e.g. at some point, I think we should make install-events more homogeneous by moving `LocalizationInitializer` under `setup/plugins/installDatabase`. But that can be another day.)

totten commented on .gitignore:
```diff
@@ -83,7 +83,6 @@ tools/tests/reports/testdox.html
 tools/tests/reports/testdox.txt
 tools/extensions/org.civicrm.contactlayout
 tools/extensions/phpstorm/.phpstorm.meta.php
-l10n
```
Suggestion: `/l10n`

That will allow it to ignore top-level `./l10n` folder while committing nested folders like `./settings/l10n`.

nganivet commented:
This is a nice improvement, I second this patch.

@yashodha  Can you please include this into our versions 5.75 and above as soon as merged upstream? Thanks.

mlutfy commented:
I did more testing and a few more tweaks:

- Moved the Localization init code to a setup plugin (it also helped fix a bug and the weird `tsLocale` global)
- Updated the text on the installer to "Language and Region", instead of Localization or Translation. Because this is equally important for the UK or Australia.
- (offtopic) allow install sample data for non-US-English. It works fine, why not?
- In the default l10n settings files, I removed the `countryLimit` and `provinceLimit` because it can be confusing, and there isn't really a performance reason to limit those anymore, unless for some reason a country lives in total isolation or has unhealthy fantasies of annexing its neighbours.
  - and side-note, for the `fr_CA` display name, I removed the  "individual prefix" and suffix, because we don't use those for communicating with people, except in very formal circumstances, and it annoys users because it's cluttery.

![image](https://github.com/user-attachments/assets/d6036977-b6b6-4e57-8a9c-1794c41423d0)

I just need to do a bit more testing because for some reason the `mo` file is not being downloaded.

mlutfy commented:
aha ok, so the mo files were not downloading because it was checking for the existence of the directory, not the mo file (and for testing, I was deleting the mo file only).

Which led me to do another change: download the mo files before launching the installation. This way, we can use the installer in the selected language (this was possible before, in the l10n.tar.gz days).

I did do an ugly hack: if mo files were downloaded, reload the page, using javascript, because digging through the controller was a bit overwhelming, and I'm not sure it would be easy to do with a `Location` header, because on various CMSes the URL would be different, and the JS hack is simple (and similar to what the dropdown does).

I'm not sure why so few strings are translated, but that's a separate issue:

![image](https://github.com/user-attachments/assets/e0ffc235-e81a-4a59-9e90-19b6463b737f)

totten commented:
> Moved the Localization init code to a setup plugin (it also helped fix a bug and the weird tsLocale global)

# :tada: :smile: 

> (offtopic) allow install sample data for non-US-English. It works fine, why not?

* Background: 
    * The sample-data still goes through a multistep process, where `regen.sh` makes a new install, then it calls some standalone PHP scripts (to fill the DB), and stores a coyp of `civicrm_generated.mysql`. That snapshot is used as the sample data.
    * The resulting data is both _random_ and _stable/reproducible_. (*If two people talk about reproducing a bug on v6.12.34 with sample-data, then they are looking at the same sample-data.*)
    * `civicrm_generated.mysql` is a *full* snapshot, including SQL strings (like "Activity Type" labels).
* Limitation:
    * Most of the runtime-UI should respond to the chosen language. However, the SQL strings in `civicrm_generated.mysql` would appear as English.
* Ideal:
    * The installer should call the data-generator directly (while using the user's chosen language). The trick is to make the output _stable/reproducible_ (e.g. random-values with a static seed-value).
    * The upshots would be:
        * Sample data is localized.
        * We don't need to commit the large `civicrm_generated.mysql` blob.
        * We never need to call `regen.sh` again. (Right now, you have to regen whenever there is a change to schema or default-data.)
        * We never get merge-conflicts on that blob. `set-version.php` can be simplified. (No need to bump the version-number in the blob.)
    * That should be feasible -- but it's probably a few hours work to get it all done+tested.

mlutfy commented:
@totten Thanks for the background, very interesting. It would be interesting indeed to get rid of regen. Could be a fun Sprint topic for me.

I reverted that part of my commit.

Random thought, while on that topic: how about removing the checkbox from the web-installer completely? I opened an issue to discuss: https://lab.civicrm.org/dev/core/-/issues/5845

totten commented:
OK, I was able to `r-run` successfully. Spot-checked a few locales and a few settings -- in each case, they initialized properly. 👍 

---

Tangentially, it took me a bit to test. I actually tried do a quick run a couple weeks ago but got side-tracked by weird problems. (*Nav-bar wouldn't render?!*) Later, the problem went away. Then I tried to test this again today... and _the problem started again_. 🤯

Happily, this patch is fine. The problem is in how I stress-test the installer. (The dev-loop should be optimized for frequent "install=>uninstall=>install=>uninstall..." cycles.). There was a separate bug with the uninstaller. After fixing that (https://github.com/civicrm/civicrm-core/pull/32661), testing this became much easier. And this works fine. 😄
